qiskit/namespace.py

```diff
@@ -0,0 +1,69 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2020.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+# pylint: disable=unused-argument
+
+"""Module for utilities to manually construct qiskit namespace"""
+
+import sys
+from importlib.abc import MetaPathFinder, Loader
+import importlib
+
+
+def _new_namespace(fullname, old_namespace, new_package):
+    names = fullname.split(".")
+    new_namespace_names = new_package.split(".")
+    old_namespace_names = old_namespace.split(".")
+    fullname = ".".join(new_namespace_names + names[len(old_namespace_names) :])
+    return fullname
+
+
+class QiskitLoader(Loader):
+    """Load qiskit element as a namespace package."""
+
+    def __init__(self, new_package, old_namespace):
+        super().__init__()
+        self.new_package = new_package
+        self.old_namespace = old_namespace
+
+    def module_repr(self, module):
+        return repr(module)
+
+    def load_module(self, fullname):
+        old_name = fullname
+        fullname = _new_namespace(fullname, self.old_namespace, self.new_package)
+        module = importlib.import_module(fullname)
+        sys.modules[old_name] = module
+        return module
+
+
+class QiskitElementImport(MetaPathFinder):
+    """Meta importer to enable unified qiskit namespace."""
+
+    def __init__(self, old_namespace, new_package):
+        super().__init__()
+        self.old_namespace = old_namespace
+        self.new_package = new_package
+
+    def find_spec(self, fullname, path=None, target=None):
+        """Return the ModuleSpec for Qiskit element."""
+        if fullname.startswith(self.old_namespace):
+            try:
+                importlib.import_module(
+                    _new_namespace(fullname, self.old_namespace, self.new_package)
+                )
+                return importlib.util.spec_from_loader(
+                    fullname, QiskitLoader(self.new_package, self.old_namespace), origin="qiskit"
+                )
+            except Exception:  # pylint: disable=broad-except
+                return None
```

Comment:
Do we need to catch `Exception` here?  Can we reduce it to `ImportError`?  I don't really know - I'm thinking of cases where there's import code error in a development version of Aer, which don't get caught because the non-import error gets masked into "package not available".

Reply:
I made this a broad exception since I didn't want a development error to break the machinery. Like I think if `import qiskit_aer` raises an `AttributeError` because of a bug I'm pretty sure this will just raise an exception and the import will fail with a weird attribute error coming from a random place. But thinking about this I think changing this to an import error is reasonable we can always change it later if needed.

Reply:
I think that's the reason I was advocating _for_ those exceptions to propagate - I think if we catch all exceptions, a development environment might accidentally pull an old version of Aer from who knows where (namespace packaging problems...) which would mask the error.  Either way, we shouldn't ever be publishing a version of Aer that errors out on `import`, so it's probably just about which way we think is better for development.

To be honest, I think I'd even go as far as limiting it to `ModuleNotFoundError` - if Aer raises an import error due to one of its compiled extensions failing to load, we'd probably want to propagate that.

Reply:
Done in https://github.com/Qiskit/qiskit-terra/pull/5089/commits/6ba893ad7c12158012e8f02e79aa3d1719abed18